### PR TITLE
[WIP] first version of example of how to split GraphModule in GraphModule with sub-GraphModules

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44910 [WIP] first version of example of how to split GraphModule in GraphModule with sub-GraphModules**

### CODE REVIEW

**Comment by Lilyjjo** [comment]
This isn't a finished version, I just wanted to see if this was in the general direction of where the issue #43813 was intended to go

File: test/test_fx.py [hunk 1]
```diff
@@ -563,6 +563,165 @@ def test_construct_root_dict(self):
...
+                    # perform arg transformation
+                    new_args = []
+                    for orig_arg in orig_node.args:
+                        if isinstance(orig_arg, torch.fx.Node):
+                            # TODO assumption that nodes aren't nested in lists/tuples
+                            new_args.append(new_nodes[orig_arg.name])
+                        else:
+                            # TODO assumption that non-nodes are directly copyable
+                            new_args.append(orig_arg)
+                    new_args = tuple(new_args)
```
Comment (jamesr66a) [inline test/test_fx.py:621]: Here and for kwargs you can use map_arg:
https://github.com/pytorch/pytorch/blob/29664e6aa3cd5b2fb9ae2dc27eecfc52f5ccd66f/torch/fx/graph.py#L55

That way it'll traverse through all the nested structures and replace `Node`s where appropriate

File: test/test_fx.py [hunk 2]
```diff
@@ -563,6 +563,165 @@ def test_construct_root_dict(self):
...
+                        target=target,
+                        args=new_args,
+                        kwargs=new_kwargs,
+                        name=orig_node.name,
+                    )
+                    if node_name == module_node_names[-1]:
+                        graph.output(new_nodes[node_name])
+
+                    if orig_node.op == "placeholder":
+                        graph_placeholders.append(new_nodes[node_name])
```
Comment (jamesr66a) [inline test/test_fx.py:650]: For the Graph inputs and outputs here, you'll want to determine which data dependencies come from outside the subgraph (inputs) and which are used outside the subgraph. I think I can explain better offline

File: test/test_fx.py [hunk 3]
```diff
@@ -563,6 +563,165 @@ def test_construct_root_dict(self):
...
+
+            # get nodes from original GraphModule
+            orig_nodes: Dict[str, torch.fx.Node] = {}
+            for node in orig_graph_module.graph.nodes:
+                orig_nodes[node.name] = node
+
+            # create subgraph A
+            subgraph_a_partition = ["x", "param", "add_1", "linear_1", "clamp_1"]
+            submodule_a, submodule_node_a = create_module_from_nodes(
+                "submod_a", subgraph_a_partition, orig_nodes, orig_torch_nn_module
```
Comment (jamesr66a) [inline test/test_fx.py:687]: Nit: using `kwargs` at the callsite of a function with many arguments (3 or more in my preference, or 2 if the arguments are easily confused) can make the code easier to understand, e.g.

```
submodule_a, submodule_node_a = create_module_from_nodes(
    module_name="submod_a", module_name_nodes=subgraph_a_partition, orig_nodes=orig_nodes, orig_base_module=orig_torch_nn_module)
```

File: test/test_fx.py [hunk 4]
```diff
@@ -563,6 +563,165 @@ def test_construct_root_dict(self):
...
+            # module_node_names: list of node names to add to new graph, need to match names in orig_nodes
+            # orig_nodes: nodes from original torch.fx.GraphModule
+            # orig_base_module: original torch.nn.module which orig_nodes's torch.fx.GraphModule was traced from
+            # submodule_hints: if adding submodules from modules created by this function, add their names here
+            #                  and add their names to module_node_names and add their nodes to orig_nodes
+            def create_module_from_nodes(
+                module_name: str,
+                module_node_names,
+                orig_nodes: Dict[str, torch.fx.Node],
+                orig_base_module: torch.nn.Module,
```
Comment (jamesr66a) [inline test/test_fx.py:598]: nit: `orig_root_module` to match up with FX's terminology

**Pull request closed without merging**